### PR TITLE
SI-7459 Handle pattern binders used as prefixes in TypeTrees.

### PR DESCRIPTION
**Pull request by retronym**
Match translation was incorrect for:

```
case t => new t.C
case D(t) => new d.C
```

We would end up with Types in TypeTrees referring to the wrong
symbols, e.g:

```
// t7459a.scala
((x0$1: this.LM) => {
  case <synthetic> val x1: this.LM = x0$1;
  case4(){
   matchEnd3(new tttt.Node[Any]())
  };
  matchEnd3(x: Any){
    x
}
```

Or:
    // t7459b.scala
     ((x0$1: CC) => {
       case <synthetic> val x1: CC = x0$1;
       case4(){
         if (x1.ne(null))
           matchEnd3(new tttt.Node[Any]())
         else
           case5()
       };

This commit:
- Changes `bindSubPats` to traverse types, as well as terms,
  in search of references to bound symbols
- Changes `Substitution` to reuse `Tree#substituteSymbols` rather
  than the home-brew substitution from `Tree`s to `Tree`s, if the
  `to` trees are all `Ident`s

I had to dance around the awkward handling of "swatches" (exception
handlers that can be implemented with JVM native type switches) by
duplicating trees to avoid seeing the results of `substituteSymbols`
in `caseDefs` after we abandon that approach if we detect the
patterns are too complex late in the game.

I also had to add an escape hatch for the "type selection from
volatile type" check in the type checker. Without this, the
translation of `pos/t7459c.scala`:

```
case <synthetic> val x1: _$1 = (null: Test.Mirror[_]).universe;
case5(){
  if (x1.isInstanceOf[Test.JavaUniverse])
    {
      <synthetic> val x2: _$1 with Test.JavaUniverse = (x1.asInstanceOf[_$1 with Test.JavaUniverse]: _$1 with Test.JavaUniverse);
      matchEnd4({
        val ju1: Test.JavaUniverse = x2;
        val f: () => x2.Type = (() => (null: x2.TypeTag[Nothing]).tpe);
```

.. triggers that error at `x2.TypeTag`.

Review by @adriaanm. I think we can do better for 2.12 as we discussed in #3490.
But we should consider booking this progress.

### CODE REVIEW

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by retronym** [comment]
Dang, the new tests I added crash in the `icode` phase with `-optimize` enabled (even with all its component flags disabled):

```
% qscalac -Ydead-code -Yinline:false -Yclosure-elim:false -Yinline-handlers:false -Yconst-opt:false test/files/run/t7459b.scala

% qscalac -optimize -Ydead-code:false -Yinline:false -Yclosure-elim:false -Yinline-handlers:false -Yconst-opt:false test/files/run/t7459b.scala
error:
  symbol value tttt does not exist in LM.<init>
     while compiling: test/files/run/t7459b.scala
        during phase: icode
     library version: version 2.11.5-20141110-000629-91e2dbf786
    compiler version: version 2.11.5-20141110-000629-91e2dbf786
  reconstructed args: -optimise

  last tree to typer: Ident(n)
       tree position: line 15 of test/files/run/t7459b.scala
            tree tpe: LM
              symbol: value n
   symbol definition: n: LM (a TermSymbol)
      symbol package: <empty>
       symbol owners: value n -> constructor CC -> class CC
           call site: class Test$delayedInit$body in package <empty>

== Source file context for tree position ==

    12   new LM().g(new CC(new LM()))
    13 }
    14 case class CC(n: LM)
    15
error: scala.reflect.internal.FatalError:
  symbol value tttt does not exist in LM.<init>
     while compiling: test/files/run/t7459b.scala
```

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -3727,8 +3727,12 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           case TypeRef(pre, sym, args) =>
             if (sym.isAliasType && containsLocal(tp) && (tp.dealias ne tp)) apply(tp.dealias)
             else {
-              if (pre.isVolatile)
-                InferTypeWithVolatileTypeSelectionError(tree, pre)
+              if (pre.isVolatile) pre match {
+                case SingleType(_, sym) if sym.isSynthetic && isPastTyper =>
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3731]: Could we push this exception (`!(isPastTyper && sym.isSynthetic) && ...`) into `isVolatile` (`... isVolatile(tp.underlying) && (sym.hasVolatileType || !sym.isStable)`)?

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:3731]: ping @retronym -- I'll merge by end of my day unless you'd like to take this on for extra credit

**Pull request merged**